File: adapters/java/src/test/java/adapter/LogicalValueTest.java
package adapter;

import static org.junit.jupiter.api.Assertions.*;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.math.BigInteger;
import java.util.List;
import org.junit.jupiter.api.Test;

class LogicalValueTest {
    private static final ObjectMapper MAPPER = new ObjectMapper()
        .configure(DeserializationFeature.USE_BIG_INTEGER_FOR_INTS, true);

    private static JsonNode parse(String json) {
        try {
            return MAPPER.readTree(json);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Test
    void parsesInt() {
        LogicalValue v = LogicalValue.parse(parse("{\"type\":\"int\",\"value\":\"42\"}"));
        assertEquals(new LogicalValue.Int("42"), v);
    }

    @Test
    void parsesFloat() {
        LogicalValue v = LogicalValue.parse(parse("{\"type\":\"float\",\"width\":\"auto\",\"value\":\"2.5\"}"));
        assertEquals(new LogicalValue.Flt("auto", "2.5"), v);
    }

    @Test
    void parsesText() {
        LogicalValue v = LogicalValue.parse(parse("{\"type\":\"text\",\"value\":\"café\"}"));
        assertEquals(new LogicalValue.Text("café"), v);
    }

    @Test
    void parsesBytes() {
        LogicalValue v = LogicalValue.parse(parse("{\"type\":\"bytes\",\"value\":\"deadbeef\"}"));
        assertEquals(new LogicalValue.Bytes("deadbeef"), v);
    }

    @Test
    void parsesBoolAndNull() {
        assertEquals(new LogicalValue.Bool(true), LogicalValue.parse(parse("{\"type\":\"bool\",\"value\":true}")));
        assertEquals(new LogicalValue.Null(), LogicalValue.parse(parse("{\"type\":\"null\"}")));
    }

    @Test
    void parsesArray() {
        LogicalValue v = LogicalValue.parse(
            parse("{\"type\":\"array\",\"items\":[{\"type\":\"int\",\"value\":\"1\"},{\"type\":\"int\",\"value\":\"2\"}]}"));
        assertEquals(new LogicalValue.Arr(List.of(new LogicalValue.Int("1"), new LogicalValue.Int("2"))), v);
    }

    @Test
    void parsesMap() {
        LogicalValue v = LogicalValue.parse(
            parse("{\"type\":\"map\",\"entries\":[[{\"type\":\"text\",\"value\":\"a\"},{\"type\":\"int\",\"value\":\"1\"}]]}"));
        assertEquals(
            new LogicalValue.Map(List.of(java.util.Map.entry(new LogicalValue.Text("a"), new LogicalValue.Int("1")))),
            v);
    }

    @Test
    void parsesTag() {
        LogicalValue v = LogicalValue.parse(parse("{\"type\":\"tag\",\"tag\":100,\"value\":{\"type\":\"int\",\"value\":\"5\"}}"));
        assertEquals(new LogicalValue.Tag(BigInteger.valueOf(100), new LogicalValue.Int("5")), v);
    }

    @Test
    void parsesTagAtU64Max() {
        // Regression: the tag field must survive as BigInteger through the
        // JSON layer without narrowing through a signed 64-bit intermediate.
        // 18446744073709551615 == 2^64-1.
        LogicalValue v = LogicalValue.parse(
            parse("{\"type\":\"tag\",\"tag\":18446744073709551615,\"value\":{\"type\":\"int\",\"value\":\"5\"}}"));
        assertEquals(new LogicalValue.Tag(new BigInteger("18446744073709551615"), new LogicalValue.Int("5")), v);
    }

    @Test
    void rejectsTagAboveU64Max() {
        assertThrows(LogicalValue.ParseException.class, () -> LogicalValue.parse(
            parse("{\"type\":\"tag\",\"tag\":18446744073709551616,\"value\":{\"type\":\"int\",\"value\":\"5\"}}")));
    }

    @Test
    void parsesBignum() {
        LogicalValue v = LogicalValue.parse(
            parse("{\"type\":\"bignum\",\"sign\":\"positive\",\"value\":\"18446744073709551616\"}"));
        assertEquals(new LogicalValue.Bignum("positive", "18446744073709551616"), v);
    }

    @Test
    void rejectsUnknownType() {
        assertThrows(LogicalValue.ParseException.class, () -> LogicalValue.parse(parse("{\"type\":\"nonsense\"}")));
    }

    @Test
    void rejectsMissingType() {
        assertThrows(LogicalValue.ParseException.class, () -> LogicalValue.parse(parse("{}")));
    }
}

File: adapters/kotlin/src/test/kotlin/LogicalValueTest.kt
import org.json.JSONObject
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class LogicalValueTest {
    @Test
    fun parsesInt() {
        val v = parseLogicalValue(JSONObject("""{"type":"int","value":"42"}"""))
        assertEquals(LogicalValue.Int("42"), v)
    }

    @Test
    fun parsesFloat() {
        val v = parseLogicalValue(JSONObject("""{"type":"float","width":"auto","value":"2.5"}"""))
        assertEquals(LogicalValue.Float("auto", "2.5"), v)
    }

    @Test
    fun parsesText() {
        val v = parseLogicalValue(JSONObject("""{"type":"text","value":"café"}"""))
        assertEquals(LogicalValue.Text("café"), v)
    }

    @Test
    fun parsesBytes() {
        val v = parseLogicalValue(JSONObject("""{"type":"bytes","value":"deadbeef"}"""))
        assertEquals(LogicalValue.Bytes("deadbeef"), v)
    }

    @Test
    fun parsesBoolAndNull() {
        assertEquals(LogicalValue.Bool(true), parseLogicalValue(JSONObject("""{"type":"bool","value":true}""")))
        assertEquals(LogicalValue.Null, parseLogicalValue(JSONObject("""{"type":"null"}""")))
    }

    @Test
    fun parsesArray() {
        val v = parseLogicalValue(
            JSONObject(
                """{"type":"array","items":[{"type":"int","value":"1"},{"type":"int","value":"2"}]}"""
            )
        )
        assertEquals(LogicalValue.Array(listOf(LogicalValue.Int("1"), LogicalValue.Int("2"))), v)
    }

    @Test
    fun parsesMap() {
        val v = parseLogicalValue(
            JSONObject(
                """{"type":"map","entries":[[{"type":"text","value":"a"},{"type":"int","value":"1"}]]}"""
            )
        )
        assertEquals(
            LogicalValue.Map(listOf(LogicalValue.Text("a") to LogicalValue.Int("1"))),
            v,
        )
    }

    @Test
    fun parsesTag() {
        val v = parseLogicalValue(
            JSONObject("""{"type":"tag","tag":100,"value":{"type":"int","value":"5"}}""")
        )
        assertEquals(LogicalValue.Tag(100uL, LogicalValue.Int("5")), v)
    }

    @Test
    fun parsesTagAtULongMax() {
        // Regression: obj.get("tag") returns a BigInteger for JSON integer
        // literals too large for Long/Int, and the old Long-based pipeline
        // silently truncated this. 18446744073709551615 == ULong.MAX_VALUE.
        val v = parseLogicalValue(
            JSONObject("""{"type":"tag","tag":18446744073709551615,"value":{"type":"int","value":"5"}}""")
        )
        assertEquals(LogicalValue.Tag(ULong.MAX_VALUE, LogicalValue.Int("5")), v)
    }

    @Test
    fun rejectsTagAboveULongMax() {
        assertFailsWith<ParseException> {
            parseLogicalValue(
                JSONObject("""{"type":"tag","tag":18446744073709551616,"value":{"type":"int","value":"5"}}""")
            )
        }
    }

    @Test
    fun parsesBignum() {
        val v = parseLogicalValue(
            JSONObject("""{"type":"bignum","sign":"positive","value":"18446744073709551616"}""")
        )
        assertEquals(LogicalValue.Bignum("positive", "18446744073709551616"), v)
    }

    @Test
    fun rejectsUnknownType() {
        assertFailsWith<ParseException> { parseLogicalValue(JSONObject("""{"type":"nonsense"}""")) }
    }

    @Test
    fun rejectsMissingType() {
        assertFailsWith<ParseException> { parseLogicalValue(JSONObject("""{}""")) }
    }
}
